### PR TITLE
add Transocoder functionality

### PR DESCRIPTION
**Pull request by Kubuxu**
Allowing multiaddr extensions to have data with

### CODE REVIEW

**Comment by lgierth** [comment]
What is the goal of this PR, and how does it achieve it? Can't review 370 changed lines with only half a sentence of description ;)

**Comment by Kubuxu** [comment]
166*2 lines of the is just code moved

Rest (50) is Transcoder interface that allows to have data in multiaddr without having your Protocol encoding and decoding inside this repo.

**Comment by lgierth** [comment]
Could you split the refactoring and added functionality into two commits?

**Comment by Kubuxu** [comment]
No it isn't possible as then commit won't build and/or pass tests.

**Comment by Kubuxu** [comment]
//cc @whyrusleeping, @lgierth for CR

File: transcoders.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+	mh "github.com/multiformats/go-multihash"
+)
+
+type Transcoder interface {
+	StringToBytes(string) ([]byte, error)
+	BytesToString([]byte) (string, error)
+}
+
+type twrp struct {
```
Comment (whyrusleeping) [inline transcoders.go]: maybe provide a function that constructs one of these given two functions?

**Comment by Kubuxu** [comment]
Updated.

**Comment by Kubuxu** [comment]
@whyrusleeping CR pls.

**Comment by whyrusleeping** [comment]
i wish git diffs were smarter. All youre really doing here is moving code around, but git thinks youre rewriting EVERYTHING.

This LGTM, but would love @lgierth to also take a look.

**Comment by lgierth** [comment]
LGTM :+1: 

It'll also simplify potentially adding multiaddr protocols with more than one value.

**Comment by whyrusleeping** [comment]
@Kubuxu coverage decreased. I'm very disapointed. (though this LGTM and i'm gonna merge it)

**Pull request merged**